React scripts 5

benjamin-heasly opened the pull request:
Hello!  This PR is in preparation for upgrading to jspsych 7 (which will be my next PR).  This one upgrades Honeycomb to use react-scripts v5 (we were on v3), which in turn gives us webpack v5.

I didn't set out to upgrade webpack! :stuck_out_tongue:  But after resisting it / being in denial for a while, I do think we require v5.

The reason is that jspsych and its plugins come in ES modules that have non-default entry points -- i.e., the module code comes from a file like `node_modules/jspsych/dist/index.js` rather than the default like `node_modules/jspsych/index.js` ([source](https://github.com/jspsych/jsPsych/blob/main/packages/jspsych/package.json#L7)).  Webpack didn't support non-default entry points like this until v5.

We could work around the entry point in our own code by importing `from 'jspsych/dist'`.  But the jspsych plugin modules also need to import jspsych, and they don't do it this way.  They directly import `from 'jspsych'` and seem to rely on the tooling like webpack to find the configured entry point for them ([source](https://github.com/jspsych/jsPsych/blob/main/packages/plugin-html-keyboard-response/src/index.ts#L1)).  So, here we go!

Webpack 5 introduces some upgrades and breaking changes, so this PR also includes various cleanup to get the build working and clean again.

The big breaking change is around polyfills for client code that tries to access modules intended for the server, for example `dotenv` ([cries of pain](https://github.com/facebook/create-react-app/issues/11756)).  We had one instance of client code trying to use `dotenv`, which in turn requires `fs`, which is no longer present for the client.  I am hoping we don't actually need to do this, and can instead rely on env vars that react-scripts exposes to us, the ones that get prefixed with `REACT_APP_`.

I'll put a few more comments in the actual change sets.

:smile_cat:

benjamin-heasly committed:
build(react-scripts): upgrade react-scripts to 5.0.1

To support jspsych 7 with esmodules, we need to upgrade our react-scripts from v3 to v5.

BREAKING CHANGE: react-scripts 5 brings webpack 5, which has breaking changes, to be addressed in
the next few commits.

re #91

benjamin-heasly committed:
fix(app version): create app-version.json that's safe to default-import in config

Webpack >3 yelled at our named import of package.json, it wants us to default import.  This creates
a minimal module app-version that is safe to default import in the app config.

re #91

benjamin-heasly committed:
fix(firebase, dotenv): remove client-side usage of dotenv from firebase.js

With webpack >=5 we can no longer access the fs module on the client side.  There seems to be no
polyfill we can add to restore this.  So removing this usage of dotenv, with the intent of relying
on REACT_APP_* variables provided by react-scripts, instead.

re #91

benjamin-heasly committed:
fix(firebase): upgrade firebase to 9.8.4

With webpack 5, firebase 8 produced many build warnings about loading source maps.  Instead of
fixing this, let's just upgrade to firebase 9, which doesn't produce the warnings.

re #91, #94

benjamin-heasly committed:
fix(eslint): get eslint config working and resolve minor lint errors

We had a few issues with the eslint config, some related to react-scripts v5 upgrade and some
already exsiting.  This is intended to get eslint running again, and in turn resolve a few minor
lint errors.

re #91

benjamin-heasly committed:
fix(bootstrap): upgrade bootstrap to 5.2.0-beta1

Building with webpack 5, bootstrap produced warnings about deprecated usage, "autoprefixer: Replace
color-adjust to print-color-adjust. The color-adjust shorthand is currently deprecated".  This
upgrades past the deprecated usage, and we can adopt 5.2.0 as soon as it becomes GA/no longer beta.

re #91

benjamin-heasly commented on .eslintrc.js:
```diff
@@ -21,5 +22,6 @@ module.exports = {
     'react'
   ],
   'rules': {
+    'react/prop-types': 'off'
```
Once eslint started running again, I got a bunch of prop-types errors.  IMO the solution with [PropTypes](https://reactjs.org/docs/typechecking-with-proptypes.html) is kinda clunky.  Since we don't have a ton of React code, I just disabled the check.  But I'm happy to revisit this.

mirestrepo replied:
I think this okay

benjamin-heasly commented on package.json:
```diff
@@ -15,12 +15,12 @@
     "@brown-ccv/behavioral-task-trials": "^1.0.0",
     "@fortawesome/fontawesome-free": "^5.9.0",
-    "bootstrap": "^4.3.1",
+    "bootstrap": "^5.2.0-beta1",
```
It's awkward to depend on a beta, but I wanted to see a clean build!  I am hoping we can just upgrade to 5.2.0 when it becomes GA / non-beta.

benjamin-heasly commented on package.json:
```diff
@@ -71,9 +71,6 @@
   },
-  "eslintConfig": {
-    "extends": "react-app"
-  },
```
I think this section was redundant with `.eslintrc.js`, and it confused me for a while, so I removed it.

benjamin-heasly commented on package.json:
```diff
@@ -98,17 +95,11 @@
-    "eslint-plugin-node": "^9.2.0",
-    "eslint-plugin-promise": "^4.2.1",
-    "eslint-plugin-react": "^7.14.3",
-    "eslint-plugin-standard": "^4.0.1",
```
These conflicted with dependencies from react-scripts v5.  Could we just accept the config and conventions from react-scripts?

hollandjg replied:
It looks like `eslint` has been removed completely as a dev dependency, but there's still an `.eslintrc`. Might upgrading to the newest `eslint` and plugins help?

benjamin-heasly replied:
We do still have eslint and plugins as dev dependencies, but we're getting them as the opinionated defaults from react-scripts instead of choosing our own.  So the `.eslintrc.js` is still being used, and I think it's working OK now.

benjamin-heasly commented on package.json:
```diff
@@ -98,17 +95,11 @@
   },
   "babel": {
-    "plugin": [
+    "plugins": [
```
This typo spammed the build with lots of confusing errors about all of our js files having an error on line `0`!  Once I added the `s`, the linter started yelling at me about actual code/style issues, so I think that was progress :smile_cat:

benjamin-heasly commented on src/config/main.js:
```diff
@@ -7,6 +7,7 @@ import _ from "lodash";
 import { eventCodes } from "./trigger";
 import { init } from "@brown-ccv/behavioral-task-trials";
 import { getProlificId } from "../lib/utils";
+import app from './app-version'
```
Grab app name and version from the `src/config/app-version.json`, which is created as a postinstall script, and gets its values out of `package.json`.

eldu replied:
If the version in packages.json gets updated, how does the `app-version.json` stay in sync?

benjamin-heasly replied:
Hi @eldu :wave: Yeah, I thought about this too and I'd be open to taking a different approach.

My thinking with this approach is that the `app-version.json` is in `.gitignore` so for a fresh checkout it won't exist until it's created by `npm install`, and at that moment it will be in sync.  I think this covers the case of builds created by automation, like GitHub Actions, so at least that important case should work out OK.

I agree it's possible for these to get out of sync during local development, though, and there's nothing at the moment to prevent that.  I think it's a low-tech, best-effort approach similar to how Honeycomb is already handling Git version info in `git-version.js` (called `version.js` before this PR).  In both cases, I think automation should be OK, but local dev could get out of sync.

Do you think we should revisit this?  I was nervous about importing `package.json` directly as a whole, but maybe this is not as bad as I thought?

eldu replied:
Importing package.json doesn't seem too bad to me especially since this repo is public anyway? I'm asking around in CCV as well and we'll see what others think! 

Could also just import what you needed from the `package.json`
`import { name, version } from "package.json"`;

benjamin-heasly replied:
OK, cool, thanks.

Thinking back, the only reason I changed this part of the code is that after the upgrade I started getting warnings like this:

> Should not import the named export 'version' (imported as 'version') from default-exporting module (only default export is available soon)

And then trying to follow up on that, it seemed like we might not want to import the whole thing, as in this [so](https://stackoverflow.com/questions/64993118/error-should-not-import-the-named-export-version-imported-as-version).

So I thought maybe another approach would be good.  But I'm also happy to revert this part and maybe open a separate issue to keep an eye on that warning?

hollandjg replied:
I'm wondering how this version parameter is actually intended to be used, as it seems a pretty complicated way to get the version number out. Is there a problem if it gets out of sync locally? perhaps storing alongside data collected from a study, to show which version of the code was used when collecting the data. Or is it not intended for that purpose? 

Is there something like this [so post](https://stackoverflow.com/a/22339262/5542581) which might be fit for purpose?

benjamin-heasly replied:
Hi @hollandjg.  Yeah, I'm not sure how the parameter is used either, but it does seem to be stored along with task data, so that would be my guess as well.

I think the env vars approach might work.  If I understand it, we'd add some lines to each of our .env files like these:
```
REACT_APP_VERSION=$npm_package_version
REACT_APP_NAME=$npm_package_name
```
And this might let us access the values via `process.env`.  But I don't know if the `npm_` variables would be set when running as an Electron app, so we might need another approach in that case, so I'm not sure if this would simplify things overall.

For now, I think I will back this change out and go with ~named imports~ a default import from package.json, like @eldu suggested and similar to what Honeycomb does already.  I'll open a separate ticket to capture the fact that with this upgrade, Webpack is starting to ~warn~ error about the named import approach, and I'll add some of what we discussed here.

This might be worth tackling on its own, and I didn't really mean to make waves with this part of the PR! :smile:

benjamin-heasly replied:
I backed out the changes with the postinstall script and went with an import from package.json.  If it's OK with you all, maybe we could decouple the app version detection from the upgrades to react-scripts and webpack?  I created https://github.com/brown-ccv/honeycomb/issues/99 to keep track of the app version detection discussion.

eldu replied:
Sounds good!

benjamin-heasly commented on package.json:
```diff
@@ -15,12 +15,12 @@
     "event-marker": "git+https://github.com/brown-ccv/event-marker.git",
     "execa": "^5.0.0",
-    "firebase": "^8.2.6",
+    "firebase": "^9.8.4",
```
Firebase v9 imported cleaner than v8 ([known issue](https://github.com/firebase/firebase-js-sdk/issues/3711)), so I just went ahead and upgraded the dependency and used "compat" imports to work with our existing code.

benjamin-heasly commented on src/firebase.js:
```diff
@@ -1,7 +1,6 @@
+import firebase from 'firebase/compat/app';
+import 'firebase/compat/firestore';
 
-require("dotenv").config();
```
With webpack v5, we are no longer able to access dotenv (and in turn fs) on the client side.  I am hoping we don't need to do this, and can instead rely on react-scripts to provide us with vars (the ones that start with REACT_APP_).

mirestrepo reviewed (commented):
This is great Ben! I’m on vacation until July 18 so I’m going to ask @hollandjg and @eldu to help providing a reviw

eldu commented on src/firebase.js:
```diff
@@ -29,17 +28,18 @@ if (window.location.hostname === "localhost") {
       return true
     })
     .catch((error) => {
+      console.log(error)
```
You could print these errors with `console.error`

```suggestion
      console.error(error)
```

benjamin-heasly replied:
Oh nice one, thanks.  I'll change it.

eldu reviewed (commented):
👋 hello! I took a look at this pr, I put in a few comments. Overall the code looks good! 

I'll approve after my question for keeping the app version in sync between the package.json and the app-version.json is answered

benjamin-heasly committed:
fix(error logging): log Firebase errors at console.error level instead of console.log

Small change to display Firebase errors at the console.error level instead of default console.log

re #91

benjamin-heasly committed:
refactor(config): remove app-verions.js script, import package.json directly instead

config/main.js was importing app name and version indirectly, via a postinstall script.  This
simplifies things to import the values from package.json directly.

re #91

eldu reviewed (approved):
This looks good! Thanks for the updates! And thanks for making the separate issue for getting the version
